test/controllers/products_controller_test.rb

```diff
@@ -0,0 +1,192 @@
+require "test_helper"
+
+describe ProductsController do
+
+  describe "index" do
+    it "allows you to see products when products exist" do
+      Product.count.must_be :>, 0, "No products"
+      get products_path
+      must_respond_with :success
+    end
+
+    it "still succceeds when there are no products" do
+      Product.destroy_all
+      get products_path
+      must_respond_with :success
+    end
+  end
+
+  describe "new" do
+    it "will allow you to see a form to create a new product" do
+      get new_product_path
+      must_respond_with :success
+    end
+  end
+
+  describe "create" do
+    let(:tanja) {users(:tanja)}
+
+    it "for logged in user, will create a new product with valid input" do
+      start_count = Product.count
+      login(tanja)
+
+
+      product_data = {
+        product: {
+        name: "Snakeskin pants",
+        categories: [categories(:clothing).id, categories(:food).id],
+        price: 999.99,
+        quantity: 2,
+        stock: "In Stock",
+        user: tanja
+      }
+    }
+
+      post products_path, params: product_data
+      new_product = Product.last
+      # puts new_product.name
+      must_redirect_to product_path(new_product.id)
+      Product.count.must_equal start_count + 1
+    end
+
+    it "will not create a product if user is not logged in" do
+      start_count = Product.count
+
+      product_data = {
+        product: {
+        name: "Snakeskin pants",
+        categories: [categories(:clothing).id, categories(:food).id],
+        price: 999.99,
+        quantity: 2,
+        stock: "In Stock",
+        user: tanja
+      }
+    }
+
+      post products_path, params: product_data
+      Product.count.must_equal start_count
+      flash[:result_text].must_equal "Error: Product was not added"
+    end
+  end
+
+  describe "show" do
+    it "will go to an existing product's page" do
+      get product_path(Product.first.id)
+      must_respond_with :success
+    end
+
+    it "will render 404 if the product does not exist" do
+      nonexistant_id = ((Product.last).id) + 1
+
+      get product_path(nonexistant_id)
+      must_respond_with :not_found
+    end
+
+  end
+
+  describe "edit" do
+    it "will get the edit form for an existing product" do
+      get edit_product_path(Product.first)
```

Comment:
What about if the user isn't logged in!!!!  OMG!  Or if the user isn't the owner!